### ISSUE TITLE
GPU Support

### ISSUE THREAD

**Issue by parano** [bentoml/BentoML#886]
**Is your feature request related to a problem? Please describe.**

Currently, in order to serve models on GPU instance with BentoML, the user is asked to manage how the GPU drivers and related libraries are installed on the production server. They will need to either manually install all dependencies and run `bentoml serve` command on the server, or use custom docker base image or the `setup_sh` option to customize the API server docker image to work with GPU.

**Describe the solution you'd like**
I would like to add a new docker base image for BentoML users to use as a default for serving with GPU.  It should make serving models trained with PyTorch, Tensorflow, FastAI.. on GPU very easy.

* Add a new Docker base image here: https://github.com/bentoml/BentoML/tree/master/docker/model-server
   * add a new dockerfile for GPU `Dockerfile-gpu`
   * add it to the release.sh script, which will let us publish GPU based images to under `bentoml/model-server:0.8.3-gpu` tag on dockerhub, similar to the `slim` images

* When saving a BentoML bundle, besides the default `Dockerfile`, we should also generate a `Dockerfile-gpu` file in the same directory. This `Dockerfile-gpu` will be using the `bentoml/model-server:0.8.3-gpu`

* The user can then choose to use `Dockerfile-gpu` to create an API server docker container for serving on GPU:
```
# go to the BentoML bundle directory:
bentoml retrieve IrisClassifier:latest --target-dir=/tmp/iris_classifier
cd /tmp/iris_classifier

# build a docker image with the gpu version dockerfile:
docker build -t iris-classifier-gpu -f ./Dockerfile-gpu .

# run the docker image with Nvidia docker runtime:
docker run --runtime=nvidia -p 5000:5000  iris-classifier-gpu
```

**Describe alternatives you've considered**

Suggestions for alternatives are welcomed! Please comment below.

**Additional context**
n/a

**Comment by jackyzha0** [comment]
this seems super cool and I would wanna try this after #716 but who knows how long that will take :joy: If anyone else is also interested, feel free to take a stab at it

**Comment by Sharathmk99** [comment]
@jackyzha0 if you have not started to implement this feature, I can take this up.

**Comment by jackyzha0** [comment]
> @jackyzha0 if you have not started to implement this feature, I can take this up.

Hey @Sharathmk99, haven't started yet. Feel free to pick it up!

**Comment by bakshienator77** [comment]
@Sharathmk99 Any chance I can use your branch of the code in case you have started working on this? Working on something time sensitive that would benefit from use of GPUs

**Comment by parano** [comment]
@bakshienator77 since it is a time sensitive project you are working on, I'd recommend trying out the custom docker base image option: https://docs.bentoml.org/en/latest/concepts.html#using-other-docker-base-images, which allows you to use your own docker base image for the API server dockerfile, and you can use a docker image that has all the GPU drivers and libraries installed.

**Comment by Sharathmk99** [comment]
@parano do we need to build docker image for multiple nvidia cuda versions like 10.1, 10.2, 11.0?

**Comment by gregd33** [comment]
It seems work on this might have stalled - I'm happy to take it up if needed.

That said, the question about which base images to build are valid. Because bentoml currently has 3 supported python versions and then potentially many supported CUDA versions (as well as cudnn), having base images for all of them would be difficult. And they all could be useful since the is much more variety in what people might have in their environment.

I'm wondering if we could potentially do this dynamically. That is, as part of the `@env` arguments, we request a `base_docker_gpu_image`.

Then, if this exists, the Dockerfile-gpu we generate is the same as he typical Dockerfile except it starts like this

```
FROM base_docker_image

FROM base_docker_gpu_image

COPY --from=0 /opt/conda /opt/conda
ENV PATH /opt/conda/bin:$PATH

```
So it will basically take the GPU base image, and copy all of the conda stuff from the base image into it. So essentially it is only built when needed and doesn't need to be released. 

The main caveat is that `base_docker_image` is conda based and not `slim`. We could check for that in the `COPY` step and raise an error then.

**Comment by parano** [comment]
@gregd33 you are welcome to contribute this!

What's a typical `base_docker_gpu_image` that a user could use? Do you think it is the user's own custom gpu base image they build from scratch? or other public images like `tensorflow/tensorflow:2.3.0-gpu`?  

Would it be helpful if we provide a few default GPU images? and in the cases where cuda or cudnn version does not match the user's requirement, they could alwasy go back to building their own GPU image as the `base_docker_image parameter` right? - would there be any problem with that approach?

**Comment by gregd33** [comment]
That is a good point. 

What I had in mind as the `base_docker_gpu_image` was a public image such as `nvidia/cuda:11.0-cudnn8-runtime-ubuntu16.04` or the other NVIDIA base images. These are basically stock OS + CUDA/cudnn. In particular, no python. So that's why the above step to add python is needed.

However, if you do use something like `tensorflow/tensorflow:2.3.0-gpu` then python here is already installed. In this case, you wouldn't really need to add anything extra (other than the python package installs that will be done by the other `env` params). That is, in this case, the image could be used exactly as the `base_docker_image` is. 

So combining those two cases, a solution might be a `Dockerfile-gpu` that looks sort of like this 

```
FROM base_docker_image

FROM base_docker_gpu_image

if python doesn't exist in environment:
  COPY --from=0 /opt/conda /opt/conda
  ENV PATH /opt/conda/bin:$PATH

(rest is the same as current Dockerfile)
```
So, if the `base_docker_gpu_image` is python-ready, then basically `Dockerfile` and `Dockerfile-gpu` will be identical other than the `FROM` line. If it is not, then we add conda in - again this relies on `base_docker_image` being a conda one.

And I could see some utility in a couple of base gpu image. In particular one for tensorflow-gpu and one for torch (https://hub.docker.com/r/pytorch/pytorch/tags). Maybe just the `latest` for each. In this case, I believe the `Dockerfile` for release would be the same as the current ones, just with a different `FROM`. It seems torch uses `conda` while tensorflow does not. 

One additional question - does anything in the bentoml codebase actually run docker containers? I ask because running with GPUs requires an additional argument (and some requirements for the docker environment)

**Comment by aarnphm** [comment]
Anyone picking on this one? I would love to work on this

**Comment by gregd33** [comment]
I put it on hold since I was not using GPUs anymore  so go for it. Basically my progress is in my last comment.

**Comment by parano** [comment]
I will take this one, @aarnphm ping me on slack if you are interested in helping!
